credentials/cache/file.go

```diff
@@ -0,0 +1,204 @@
+package cache
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"golang.org/x/oauth2"
+)
+
+const (
+	// tokenCacheFile is the path of the default token cache, relative to the
+	// user's home directory.
+	tokenCacheFilePath = ".databricks/token-cache.json"
+
+	// ownerExecReadWrite is the permission for the .databricks directory.
+	ownerExecReadWrite = 0o700
+
+	// ownerReadWrite is the permission for the token-cache.json file.
+	ownerReadWrite = 0o600
+
+	// tokenCacheVersion is the version of the token cache file format.
+	//
+	// Version 1 format:
+	//
+	// {
+	//   "version": 1,
+	//   "tokens": {
+	//     "<key>": {
+	//       "access_token": "<access_token>",
+	//       "token_type": "<token_type>",
+	//       "refresh_token": "<refresh_token>",
+	//       "expiry": "<expiry>"
+	//     }
+	//   }
+	// }
+	tokenCacheVersion = 1
+
+	// lockFilePath is the path of the lock file used to prevent concurrent
+	// reads and writes to the token cache file.
+	lockFilePath = ".databricks/token-cache.lock"
+)
+
+// The format of the token cache file.
+type tokenCacheFile struct {
+	Version int                      `json:"version"`
+	Tokens  map[string]*oauth2.Token `json:"tokens"`
+}
+
+type FileTokenCacheOpt func(*FileTokenCache)
+
+func WithFileLocation(fileLocation string) FileTokenCacheOpt {
+	return func(c *FileTokenCache) {
+		c.fileLocation = fileLocation
+	}
+}
+
+func WithLocker(locker sync.Locker) FileTokenCacheOpt {
+	return func(c *FileTokenCache) {
+		c.locker = locker
+	}
+}
+
+// FileTokenCache caches tokens in "~/.databricks/token-cache.json". FileTokenCache
+// implements the TokenCache interface.
+type FileTokenCache struct {
+	fileLocation string
+
+	// locker protects the token cache file from concurrent reads and writes.
+	locker sync.Locker
+}
+
+// NewFileTokenCache creates a new FileTokenCache. By default, the cache is
+// stored in "~/.databricks/token-cache.json". The cache file is created if it
+// does not already exist. The cache file is created with owner permissions
+// 0600 and the directory is created with owner permissions 0700. If the cache
+// file is corrupt or if its version does not match tokenCacheVersion, an error
+// is returned.
+func NewFileTokenCache(opts ...FileTokenCacheOpt) (*FileTokenCache, error) {
```

Comment:
I think we should return a `TokenCache` here rather than a `FileTokenCache`. The reasons are:

1. we do not want users to create a `FileTokenCache` without using this constructor,
2. the content of the FileTokenCache is private anyway, and
3. `FileTokenCache` does not expose new methods compared to the interface it implements.  

```suggestion
func NewFileTokenCache(opts ...FileTokenCacheOpt) (TokenCache, error) {
```